### DISCUSSION TITLE
What are Cross-process Islands?

### DISCUSSION THREAD

**Discussion by Marv51** [microsoft/WindowsAppSDK#3892] [Q&A]
If I understand correctly the current XAML-Islands enable devs to embed WinUI 3 islands in any other UI framework.

What are the use-cases for the "multi-release effort to support Cross-process Islands, including fundamental input and rendering support for Content Islands"? This is from the [roadmap](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md)

I have always found having multiple processes in my app rather cumbersome, because of the IPC needed.

**Comment by codendone** [comment]
Cross-process islands are to support an island running in process A hosted in process B. This can provide security and reliability wins, similar to how WebView2 content runs in a separate process which doesn't have access to data in the host process and doesn't crash the host process if the WebView2 process crashes.

**Reply by alfred-png** [reply]
For now, the only cross-process island I know is the WebView2 control. Are there other controls (planned to get) available? If yes, what are they for?
Thank you in advance.

**Reply by codendone** [reply]
Cross-process islands aren't specific to any XAML controls. As another example, imagine you wanted to provide an API for downloading images from an online account. This API might not trust the calling app with the account credentials, so a secure approach could be to host a cross-process island which securely handles account login and image selection and then only provides those selected images to the calling app. The cross-process island in this case could include multiple controls for account login plus a view which displays the available images and the selected images. From the calling app's perspective, this cross-process island still feels like a native view in the app while the island actually runs securely in a separate process.

**Reply by alfred-png** [reply]
Thank you for your answer. So, if I understand you correctly, in the future, in WinUI 3 (e.g. C++/WinRT unpackaged desktop project), it's possible to create a custom API running in a cross-process island that will have its separate UI thread? If yes, will it look like a Windows Runtime Component that one could implement e.g. in C++/WinRT?

**Reply by codendone** [reply]
Correct. The API shape still needs to be determined. Initially this will likely require the app to manually launch its known island process(es) and execute some custom handshakes to hook things up (which will still enable the security and reliability advantages of cross-proc islands). But, as you suggested, ideally there will be a more convenient mechanism in the future where these components can be registered and easily activated and connected into the app.

**Reply by alfred-png** [reply]
Thanks. It's nice to get immediate replies. When can we start creating custom cross-process islands? What is the timeline?

**Reply by codendone** [reply]
The exact timeframe isn't certain yet. A lot of foundation was completed in 1.5, and there will be more in 1.6, probably enough to support cross-thread islands but not yet cross-process islands.

(Cross-thread islands could still be valuable to separate an expensive control from the rest of the main UI thread. For example, if you had a control which ran an expensive 3D rendering process, or if you had a control which frequently made blocking calls of some type [disk, network, whatever]. Moving either of those off thread would ensure the rest of the UI remained responsive even if these off-thread controls were a bit laggy.)

**Reply by amtopel** [reply]
Are cross-process islands still in the works?  And if so, can the island be hosted within a WinUI element in the client process, or just within an HWND?  Thanks for any info.

**Reply by nikolayvpavlov** [reply]
This is a very intriguing idea, but please don't re-invent ActiveX / OLE.  Let's keep it simple, and it is fine that the two processes use some established ways of talking with each other: messages, pipes, even standard input/output is fine, and COM is already there.  

What will be very appreciated is to add message-handling to Window class, similar to WinForms.WndMessage protected virtual method.  I know we can achieve it with tools like WinUIEx, or by subclassing the HWND, but we must not be forced to go to deep Win32 for simplest of cases.
